WordPress/Classes/Services/MediaService.m

```diff
@@ -16,140 +16,8 @@
-                completion(nil, error);
-            }
-            return;
-        }
-        
-        if (blog.allowedFileTypes != nil) {
-            // HEIC isn't supported when uploading an image, so we filter it out (http://git.io/JJAae)
-            NSMutableSet *mutableAllowedFileTypes = [blog.allowedFileTypes mutableCopy];
-            [mutableAllowedFileTypes removeObject:@"heic"];
-            allowedFileTypes = mutableAllowedFileTypes;
-        }
-
-        if (post != nil) {
-            media = [Media makeMediaWithPost:post];
-        } else {
-            media = [Media makeMediaWithBlog:blog];
-        }
-        media.mediaType = exportable.assetMediaType;
-        media.remoteStatus = MediaRemoteStatusProcessing;
-
-        [self.managedObjectContext obtainPermanentIDsForObjects:@[media] error:nil];
-        [[ContextManager sharedInstance] saveContextAndWait:self.managedObjectContext];
-    }];
-    if (media == nil) {
-        return nil;
-    }
-    NSManagedObjectID *mediaObjectID = media.objectID;
-    [self.managedObjectContext performBlock:^{
-        // Setup completion handlers
-        void(^completionWithMedia)(Media *) = ^(Media *media) {
-            media.remoteStatus = MediaRemoteStatusLocal;
-            media.error = nil;
-            // Pre-generate a thumbnail image, see the method notes.
-            [self exportPlaceholderThumbnailForMedia:media
-                                          completion:^(NSURL *url){
-                                              if (thumbnailCallback) {
-                                                  thumbnailCallback(media, url);
-                                              }
-                                          }];
-            if (completion) {
-                completion(media, nil);
-            }
-        };
-        void(^completionWithError)( NSError *) = ^(NSError *error) {
-            Media *mediaInContext = (Media *)[self.managedObjectContext existingObjectWithID:mediaObjectID error:nil];
-            if (mediaInContext) {
-                mediaInContext.error = error;
-                mediaInContext.remoteStatus = MediaRemoteStatusFailed;
-                [[ContextManager sharedInstance] saveContext:self.managedObjectContext];
-            }
-            if (completion) {
-                completion(media, error);
-            }
-        };
-
-        // Export based on the type of the exportable.
-        MediaImportService *importService = [[MediaImportService alloc] initWithManagedObjectContext:self.managedObjectContext];
-        importService.allowableFileExtensions = allowedFileTypes;
-        NSProgress *importProgress = [importService importResource:exportable toMedia:media onCompletion:completionWithMedia onError:completionWithError];
-        [createProgress addChild:importProgress withPendingUnitCount:1];
-    }];
-    if (progress != nil) {
-        *progress = createProgress;
-    }
-    return media;
-}
```

Comment:
This function is moved to `MediaImportService` and duplicated as two different variants: one that returns a `Media` instance which can only be called from the main thread, and the other that can be called from any thread.